### PR TITLE
Fix the deadlock caused by the ImagePullDeleteLock

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Fix a dead lock that was caused by the ImageCleanup and Image Pull. Issue #833 

### Implementation details

Based on the explanation [here](https://github.com/aws/amazon-ecs-agent/issues/833#issuecomment-307194186), adjust the order of lock acquiring in image cleanup, so that both go routines will require the lock in the same order, which wouldn't cause deadlock.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Description for the changelog

Fix an issue which could potentially cause agent running into deadlock.
### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

**Comment by vsiddharth** [comment]
These changes should eliminate the deadlock we encountered with #833 
LGTM

File: agent/engine/docker_image_manager.go [hunk 1]
```diff
@@ -275,7 +276,12 @@ func (imageManager *dockerImageManager) performPeriodicImageCleanup(ctx context.
 }
 
 func (imageManager *dockerImageManager) removeUnusedImages() {
-	seelog.Infof("Begin building map of eligible unused images for deletion")
+	seelog.Debug("Attempting to obtain ImagePullDeleteLock for removing images")
+	ImagePullDeleteLock.Lock()
+	seelog.Debug("Obtained ImagePullDeleteLock for removing images")
+	defer seelog.Debug("Released ImagePullDeleteLock after removing images")
```
Comment (petderek) [inline agent/engine/docker_image_manager.go]: Maybe group these statements in an anonymous function so its easier to see the order of execution?
```
defer func() { ... }
```

File: agent/engine/docker_image_manager_test.go [hunk 2]
```diff
@@ -34,6 +34,49 @@ import (
...
+		Image: "busybox",
+	}
+	sleepContainerImageInspected := &docker.Image{
+		ID: "sha256:qwerty",
+	}
+
+	// Cause a fake delay when recording container reference so that the
+	// race condition between ImagePullLock and updateLock gets exercised
+	// If updateLock precedes ImagePullLock, it can cause a deadlock
+	client.EXPECT().InspectImage(sleepContainer.Image).Do(func(image string) {
```
Comment (petderek) [inline agent/engine/docker_image_manager_test.go]: Cool! Did we verify that this fails before the patch?

Comment (richardpen) [inline agent/engine/docker_image_manager_test.go]: Yes, this test doesn't pass without the patch.

**Pull request merged**